magma/interface.py

```diff
@@ -234,3 +262,34 @@ def DeclareInterface(*decl, **kwargs):
     name = f"Interface({', '.join([str(a) for a in decl])})"
     dct = dict(Decl=decl, **kwargs)
     return InterfaceKind(name, (_DeclareInterface,), dct)
+
+
+def DeclareLazyInterface(io, **kwargs):
+    """LazyInterface factory function"""
+    decl = io.decl()
+    name = f"Interface({', '.join([str(a) for a in decl])})"
+    dct = dict(io=io, Decl=io.decl(), **kwargs)
+    return InterfaceKind(name, (_DeclareLazyInterface,), dct)
+
+
+class IO:
+    """
+    Class for creating an interface bundle.
+
+    @kwargs: ordered dict of {name: type}, ala decl.
+    """
+    # Note that because we require kwargs to be ordered, we have a strong
+    # requirement here for >= python version 3.6. See
+    # https://www.python.org/dev/peps/pep-0468/.
+    def __init__(self, **kwargs):
+        self.ports = {}
+        self.__decl = []
+        for name, typ in kwargs.items():
+            ref = LazyDefnRef(name=name)
+            port = typ.flip()(name=ref)
```

Comment:
We could insert a type check here on `name, type` to be `str, subclass(Type)` to catch errors early

Reply:
Agreed, though after several conversations w/ @cdonovick, he's convinced me that kind of error checking is not pythonic 😄but might be misinterpreting his point

Reply:
@rsetaluri its not pythonic to type check because duck typing.

Reply:
Now sometimes it's type checking will give you significantly better error messages.  But in those cases I would promote checking an abstract base class not the existing implementation.